Add HostNetworking container option to API.

vmarmol opened the pull request:
This allows a container to run within the same networking namespace as
the host. This will be locked down by default using a flag on the master
and nodes (similar to how privileged is handled today).

vmarmol commented:
/cc @bgrant0607 @dchen1107 @thockin @ArtfulCoder

dchen1107 commented:
LGTM

dchen1107 commented:
Waiting for @bgrant0607's lg :-)

bgrant0607 commented on pkg/api/v1beta3/types.go:
```diff
@@ -586,6 +586,8 @@ type PodSpec struct {
 	// resource requirements.
 	Host string `json:"host,omitempty" description:"host requested for this pod"`
+	// Optional: Default to false.
+	HostNetworking bool `json:"host_networking,omitempty description:"whether or not the container runs in the network namespace of the host; defaults to false; cannot be updated""`
```
description: It affects the whole pod, not just an individual container.

json: We use camelCase, not underscores.

I suggest calling it `HostNetwork`/`hostNetwork` (drop the -ing).

Have you thought about the impact on any specified container ports?
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/api/v1beta3/types.go#L296

Should we ban them if hostNetwork is true? Or should we require hostPort == containerPort (which we could/would also set by default)?

vmarmol replied:
Done and added the defaulting and validation you specified. PTAL.

vmarmol committed:
Add HostNetworking container option to API.

This allows a container to run within the same networking namespace as
the host. This will be locked down by default using a flag on the master
and nodes (similar to how privileged is handled today).

bgrant0607 commented on pkg/api/v1beta1/defaults.go:
```diff
@@ -102,3 +108,14 @@ func init() {
+	for i := range *containers {
+		for j := range (*containers)[i].Ports {
+			if (*containers)[i].Ports[j].ContainerPort == 0 {
+				(*containers)[i].Ports[j].ContainerPort = (*containers)[i].Ports[j].HostPort
```
Sorry I wasn't clear. Look at types.go. omitempty implies optional. ContainerPort is required. HostPort is optional. HostPort should be defaulted from ContainerPort.

ArtfulCoder replied:
We used HostPort, because user has to specify what HostIP, they are using with the Port.
They wont specify ContainerIP.
Since HostPort goes well with HostIP, we used that instead of using ContainerIP, though both HostPort and ContainerPort are really the same in the net=host case.

bgrant0607 replied:
@ArtfulCoder Sorry, I'm confused. There is no ContainerIP field.

ContainerPort is required, so it cannot be defaulted.

ArtfulCoder replied:
We changed it so that ContainerPort is not required if HostNetwork=true.

What I am trying to say is that, when a user defines ports in the HostNetwork=true scenario,
they still need to define which HostIP they plan to use. So, since they are specifying HostIP, i thought it was reasonable to expect that they specify HostPort as well.
ContainerPort can be considered optional in this case because, there is no container network namespace.

thockin replied:
I prefer the simpler semantic - ContainerPort is always required (in as
much as the whole structure is optional), and iff HostNetwork then default
HostPort = ContainerPort.  Have to handle the case of a user setting
hostport != containerport.

On Mon, Mar 23, 2015 at 11:43 PM, Abhi Shah <notifications@github.com>
wrote:

> In pkg/api/v1beta1/defaults.go
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/5814#discussion_r27005890>
> :
>
> > @@ -102,3 +108,14 @@ func init() {
> >  		},
> >  	)
> >  }
> > +
> > +// With host networking default all container ports to host ports.
> > +func defaultHostNetworkPorts(containers *[]Container) {
> > +	for i := range *containers {
> > +		for j := range (*containers)[i].Ports {
> > +			if (*containers)[i].Ports[j].ContainerPort == 0 {
> > +				(*containers)[i].Ports[j].ContainerPort = (*containers)[i].Ports[j].HostPort
>
> We changed it so that ContainerPort is not required if HostNetwork=true.
>
> What I am trying to say is that, when a user defines ports in the
> HostNetwork=true scenario,
> they still need to define which HostIP they plan to use. So, since they
> are specifying HostIP, i thought it was reasonable to expect that they
> specify HostPort as well.
> ContainerPort can be considered optional in this case because, there is no
> container network namespace.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/5814/files#r27005890>
> .
>

ArtfulCoder replied:
Sounds good. We handle the case of hostport !=containerport when we do validation.

dchen1107 replied:
+1 on what @thockin's semantic

vmarmol replied:
Done.

bgrant0607 commented on pkg/api/v1beta1/types.go:
```diff
@@ -1087,6 +1089,8 @@ type PodSpec struct {
 	// resource requirements.
 	Host string `json:"host,omitempty" description:"host requested for this pod"`
+	// Optional: Default to false.
+	HostNetwork bool `json:"hostNetwork,omitempty description:"whether or not the containerc in the pod run in the network namespace of the host; defaults to false; cannot be updated""`
```
Typo. Just copy the description from the previous instance.

vmarmol replied:
Done. Thanks for catching.

thockin commented on pkg/api/types.go:
```diff
@@ -587,6 +587,8 @@ type PodSpec struct {
 	// the the scheduler simply schedules this pod onto that host, assuming that it fits
 	// resource requirements.
 	Host string `json:"host,omitempty"`
+	// Optional: Default to false.
```
Comment on what this does.  Comment that anyone who sets this really should make sure to declare their ports.

vmarmol replied:
Done.

thockin commented:
LGTM, nits

bgrant0607 commented:
Please post a comment once this PR is updated to address feedback.

vmarmol commented:
All changes made, PTAL @bgrant0607

bgrant0607 commented on pkg/api/v1beta3/types.go:
```diff
@@ -586,6 +587,10 @@ type PodSpec struct {
+	// Uses the host's network namespace. If this option is set, the ports that will be
+	// used must be specified.
+	// Optional: Default to false.
+	HostNetwork bool `json:"hostNetwork,omitempty"`
```
Needs description

vmarmol replied:
Done

bgrant0607 commented on pkg/api/v1beta2/types.go:
```diff
@@ -1109,6 +1110,10 @@ type ContainerManifest struct {
+	// Uses the host's network namespace. If this option is set, the ports that will be
+	// used must be specified.
+	// Optional: Default to false.
+	HostNetwork bool `json:"hostNetwork,omitempty"`
```
Needs description

vmarmol replied:
Done

bgrant0607 commented on pkg/api/v1beta2/types.go:
```diff
@@ -1149,6 +1154,10 @@ type PodSpec struct {
+	// Uses the host's network namespace. If this option is set, the ports that will be
+	// used must be specified.
+	// Optional: Default to false.
+	HostNetwork bool `json:"hostNetwork,omitempty"`
```
Needs description

vmarmol replied:
Done

bgrant0607 commented on pkg/api/v1beta1/types.go:
```diff
@@ -1087,6 +1092,10 @@ type PodSpec struct {
+	// Uses the host's network namespace. If this option is set, the ports that will be
+	// used must be specified.
+	// Optional: Default to false.
+	HostNetwork bool `json:"hostNetwork,omitempty"`
```
Needs description

vmarmol replied:
Done

bgrant0607 commented on pkg/api/v1beta1/types.go:
```diff
@@ -63,6 +63,10 @@ type ContainerManifest struct {
+	// Uses the host's network namespace. If this option is set, the ports that will be
+	// used must be specified.
+	// Optional: Default to false.
+	HostNetwork bool `json:"hostNetwork,omitempty"`
```
Needs description

vmarmol replied:
Done

bgrant0607 commented:
Do you have pre-submit hooks installed? They should have prevented you from committing this change.

vmarmol commented:
@bgrant0607 I did not, but I just added them. Thanks for the pointer!

vmarmol commented:
Do we have a test that covers this in the CI (to catch people like me who may have forgotten the hooks?)

bgrant0607 commented:
LGTM

vmarmol commented:
Thanks for the quick turn-around @bgrant0607!

bgrant0607 commented:
hack/verify-description.sh is run by .travis.yml, but has been broken before. Shippable passed, so clearly it's not working, at least on Shippable.
